chore(lightning): Upgrade To Blocktank v2 API

coreyphillips opened the pull request:
### Description

Upgrades Bitkit to the Blocktank v2 API [here](https://github.com/synonymdev/blocktank-lsp-http/tree/master/client).

coreyphillips commented:
@limpbrains, any thoughts on why this test might be failing?

limpbrains commented:
test expected QuickSetupReserveNote to be shown, but in reality QuickSetupBlocktankNote is shown. I've changed test to reflect it.
I've removed all sleep(1000), it is really not needed, when you run test with E2E=true.
Also test was failing with `[BlocktankClientError: Failed to create order. "Channel size too big. Max size is USD999."]` on the last step when buying big barrel on CustomSetup screen. Looks like a rounding bug in Bitkit ot Blocktank. I changed test to use medium barrel - $500. Maybe this needs to be fixed and reverted?

Here is a video of the test run:

https://github.com/synonymdev/bitkit/assets/155891/b06fd547-6c41-48f9-9863-54c92b1d3f08

coreyphillips commented:
Looks like the total channel size (incoming + receiving) was exceeding the `maxChannelSizeSat` value when the user opted for a spending value + the max receiving value in the custom flow. So in [this commit](https://github.com/synonymdev/bitkit/pull/1216/commits/21ab801752a851302976b5cf7e19e5dc70eb1f68) we subtract any spending amount from the max receiving to ensure it remains under that limit.

coreyphillips commented:
I no longer understand how to appease the testing gods as @limpbrains did, but it's at least ready for review. @pwltr

limpbrains commented:
Slider on the quick setup screen feels broken

https://github.com/synonymdev/bitkit/assets/155891/b8a2e7e2-0398-4e2e-811f-8a665ccf319e

limpbrains commented:
Testing Gods are Pleased

pwltr commented:
Bitkti UI is showing channel as 'paid' on home screen even though payment was not confirmed with BT yet. On the ChannelDetails screen shows the correct state.

```
{"id":"30798b6f-eabc-49f3-a1e3-a32a5cec7d07","state":"created","feeSat":193730,"lspBalanceSat":194564,"clientBalanceSat":192638,"zeroConf":false,"channelExpiryWeeks":12,"channelExpiresAt":"2023-11-20T10:54:42.561Z","orderExpiresAt":"2023-08-28T10:54:42.561Z","channel":null,"lspNode":{"alias":"Synonym-Regtest-1","pubkey":"03b9a456fb45d5ac98c02040d39aec77fa3eeb41fd22cf40b862b393bcfc43473a","connectionStrings":["03b9a456fb45d5ac98c02040d39aec77fa3eeb41fd22cf40b862b393bcfc43473a@35.233.47.252:9400"]},"lnurl":"LNURL1DP68GUP69UHKCMMRV9KXSMMNWSARJVPSXQHKZURF9AKZ7VESXUUNSC3KVCKK2CTZVVKNGWTXXVKKZVT9XVKKZVEJVY6KXETRXAJRQDCVA7KYJ","payment":{"state":"created","paidSat":0,"bolt11Invoice":{"request":"lnbcrt1937300n1pjwca02pp5xfc4ltue0jsn3w74fch6frfsyevewfn845defgkxargryw3t7g7sdzzgfkx7cmtw3sku6eqgd5xzmnwv4kzqge4gdq4gnn4gemkynrr23x85cnswf54y7fcx5cqzzsxqrrsssp59f4g0xedn5e7rfhxl0302r6mqls7kj9yya43hvnlztqwltclj78q9qyyssq7s4cmzzq79hngyf693vgfv6a767fp75wz6gv50s74np0egmr603za5kwusdzwp4d7ppgdjnwvh44xa24eutyy72w58l378gw5m9ev0qqxt6j5u","state":"pending","expiresAt":"2023-08-28T11:24:42.000Z","updatedAt":"2023-08-28T10:24:42.000Z"},"onchain":{"address":"bcrt1qld3dg32d2ual9whsz9y3uf4nj3qtadcz2hdjvf","confirmedSat":0,"transactions":[{"amountSat":193730,"txId":"9d33bc46dc65f627381eeceab0cea67984ee209b0e6d5e1f9d1a55116291b268","vout":1,"blockConfirmationCount":0,"feeRateSatPerVbyte":11.03914590747331,"confirmed":false}]}},"couponCode":"","discountPercent":0,"updatedAt":"2023-08-28T10:24:54.067Z","createdAt":"2023-08-28T10:24:42.561Z"}
```

![Simulator Screenshot - iPhone 14 - 2023-08-28 at 12 26 49](https://github.com/synonymdev/bitkit/assets/8538369/28a21e1c-c530-46ba-a88f-f534c58e7b9b)
![Simulator Screenshot - iPhone 14 - 2023-08-28 at 12 26 47](https://github.com/synonymdev/bitkit/assets/8538369/abe72f39-3b5f-4d7f-8621-ed8aca7b5dad)
![Simulator Screenshot - iPhone 14 - 2023-08-28 at 12 26 54](https://github.com/synonymdev/bitkit/assets/8538369/839a2d3d-3fbb-464e-bcd1-eb3571396d62)

pwltr commented on src/store/actions/blocktank.ts:
```diff
@@ -367,30 +336,31 @@ export const addPaidBlocktankOrder = ({
 
 	// opening connection
-	if (order.state === 300) {
+	if (order.channel?.state === BtOpenChannelState.OPENING) {
```
`[TypeError: Cannot read property 'OPENING' of undefined]`

It would make sense to remove `declare` keyword from `BtOpenChannelState` so the enum is compiled to JS to be usable.

coreyphillips replied:
Set them as strings for now in this [commit](https://github.com/synonymdev/bitkit/pull/1216/commits/0c58170fac0d390f87b2b14bc99c7eaae4d22196) until the PR above is merged.

coreyphillips replied:
Set them as string for now in this [commit](https://github.com/synonymdev/bitkit/pull/1216/commits/0c58170fac0d390f87b2b14bc99c7eaae4d22196) until the PR above is merged.

pwltr commented:
- All of these are missing now:
<img width="924" alt="Screenshot 2023-08-30 at 15 05 30" src="https://github.com/synonymdev/bitkit/assets/8538369/e58cea60-fc1a-490c-9fc9-803dc30cd2b5">


- I have to manually pull to refresh for the channel to be opened

- After the channel was opened the ChannelDetails view still shows the wrong status "Opening connection".

- Wrong fee in transfer flow:
<img width="33%" alt="Screenshot 2023-08-30 at 15 05 30" src="https://github.com/synonymdev/bitkit/assets/8538369/21955b03-074c-40f2-8717-882f563cdd37">

- Orange bar overshoots the marker:
<img width="33%" alt="Screenshot 2023-08-30 at 15 05 30" src="https://github.com/synonymdev/bitkit/assets/8538369/071f5d91-a63d-437f-b122-6c7d827cdc93">

- Getting the receive modal when closing channels

- After the channel was closed the ChannelDetails view still shows the wrong status.

coreyphillips commented:
As of the latest commit:
- The cards/notification should be restored when setting up and opening a channel.
- `ChannelDetails` should show the correct status.

I was not able to reproduce:
- I have to manually pull to refresh for the channel to be opened.
- Wrong fee in transfer flow.

"Getting the receive modal when closing channels" was always present as well. We can address that in a separate PR.

pwltr commented:
> Wrong fee in transfer flow.

The way to reproduce this is to open 2 channels: first one for 20% spending and then a transfer for 80% spending balance.

> I have to manually pull to refresh for the channel to be opened.

Getting this behaviour consistently. This time I had to restart the app for the channel to be opened. There is a call to `restartLdk()` in `actions/blocktank.ts`, maybe that's not running?

> Getting the receive modal when closing channels" was always present as well. We can address that in a separate PR.

There is logic [here](https://github.com/synonymdev/bitkit/blob/8aa9a2802e37980c8c8c86ad01ab50f5c146f3a9/src/store/actions/wallet.ts#L1206C4-L1206C4) that should prevent that. Probably has to be update for your PR?

- I'm getting an empty toast message on channel open.
<img src="https://github.com/synonymdev/bitkit/assets/8538369/2266eddc-037d-40bb-942f-93c00cd8810a" width="33%" />

- The cards for "connecting" and "ready" should only be shown for the initial connection (first channel), all subsequent channels should only show the "transfer to spending" card. See Figma "Transfer Funds to Savings (LN channel closing)"

coreyphillips commented:
In the latest set of commits I've fixed the empty toast message and the connecting/ready cards. I've also swapped out the `restartLdk` method with the `refreshLdk` method in blocktank's `handleOrderStateChange` so that should help resolve the issue with channel opens.

Unfortunately, I'm still not able to reproduce the wrong fee bug. 

Regarding the logic [here](https://github.com/synonymdev/bitkit/blob/8aa9a2802e37980c8c8c86ad01ab50f5c146f3a9/src/store/actions/wallet.ts#L1206C4-L1206C4). The reason this never worked is because the closing txid between Blocktank and what Bitkit ends up seeing are not the same when closing a channel. LDK broadcasts a separate tx to the receiving address provided by Bitkit. So that will need some additional work to get sorted.

pwltr commented:
> In the latest set of commits I've fixed the empty toast message and the connecting/ready cards. I've also swapped out the restartLdk method with the refreshLdk method in blocktank's handleOrderStateChange so that should help resolve the issue with channel opens.

Seems to be better now.

> Regarding the logic [here](https://github.com/synonymdev/bitkit/blob/8aa9a2802e37980c8c8c86ad01ab50f5c146f3a9/src/store/actions/wallet.ts#L1206C4-L1206C4). The reason this never worked is because the closing txid between Blocktank and what Bitkit ends up seeing are not the same when closing a channel. LDK broadcasts a separate tx to the receiving address provided by Bitkit. So that will need some additional work to get sorted.

Follow up issue [here](https://github.com/synonymdev/bitkit/issues/1221)

pwltr commented:
The custom setup flow is broken when selecting the maximum local balance:

https://github.com/synonymdev/bitkit/assets/8538369/24d08f14-430c-4ae1-bbb4-3541bf626f36

coreyphillips commented on src/screens/Transfer/Confirm.tsx:
```diff
@@ -47,17 +47,27 @@ const Confirm = ({
-	const blocktankPurchaseFee = useDisplayValues(order?.price ?? 0);
+	const feeSat = order?.feeSat ?? 0;
+	const clientBalanceSat = order?.clientBalanceSat ?? 0;
+	const blocktankPurchaseFee = useDisplayValues(feeSat + clientBalanceSat);
```
Updated [here](https://github.com/synonymdev/bitkit/pull/1216/commits/0eeed2f00cd549e0917c9d674552e8f146f0b142).

coreyphillips commented:
Updated the calculations in [this commit](https://github.com/synonymdev/bitkit/pull/1216/commits/0eeed2f00cd549e0917c9d674552e8f146f0b142). Custom channels should get calculated correctly now no matter how many sats the user has.

pwltr commented:
I opened a channel with 80% spending balance and my receiving capacity is very small. I believe BT should always give me a balanced channel.

![Simulator Screenshot - iPhone 14 - 2023-09-01 at 12 50 10](https://github.com/synonymdev/bitkit/assets/8538369/c49904a7-3733-45e3-82da-8956f848a10e)

coreyphillips commented:
I'm surprised Blocktank allowed that channel purchase. The new api usually does not allow the client balance to exceed the lsp balance with the following error: `"clientBalanceSat is too big compared to lspBalanceSat. Max ratio: 1."`. 

In the latest commit, I've updated the `QuickSetup` component to set the max value on load to prevent the issue you posted above. I also adjusted the `Cost:` value to reflect the fee instead of the total sats needed to send to Blocktank.

Should be good to go for testing, just need to update the test script so that it passes with the new values for lightning onboarding.

coreyphillips commented:
We've made it pretty far in this PR. I think we should call a checkpoint here so we can get some additional testing in with the new api and apply needed fixes in separate PR's from there

The Cost is the Blocktank fee. I don't believe we're able to calculate the cost of the transaction at this point. We might want to change the verbiage to "Blocktank Fee" or something similar.

The 80% is part of a larger discussion especially as anchor channels emerge.

coreyphillips commented:
The fee Cost discrepancy should be sorted as of the latest set of commits.

coreyphillips commented:
Also increased the lsp balance when setting 0 sats on the clients end to something more usable. Should be good to go.

pwltr commented:
Channel balancing seems better now, it's still defaulting to 80% spending balance in QuickSetup though. From my tests that doesn't affect balancing at all, and it's not in the design this way, so this should be reverted. Also there are a lot of leftover console logs.

coreyphillips commented:
@pwltr Could you expand on this? Do you mean it's defaulting to a value greater than $1000 at 80%?
"It's still defaulting to 80% spending balance in QuickSetup though. From my tests that doesn't affect balancing at all, and it's not in the design this way"

pwltr commented:
I'm talking about the slider on the QuickSetup screen, you've changed it to be set at 80% initially for a reason that is not apparent to me. It should be set to 20%.

![image](https://github.com/synonymdev/bitkit/assets/8538369/b5349f9c-35a1-45e3-a4c3-294dba5883e6)

coreyphillips commented:
Gotcha. Updated [here](https://github.com/synonymdev/bitkit/pull/1216/commits/8f05e630082b34952e8ffbbbc53c32b114f0494a). Changed it from setting it to the max of 80% to 20%.

pwltr commented on src/store/actions/blocktank.ts:
```diff
@@ -265,19 +258,33 @@ export const startChannelPurchase = async ({
 	});
 
-	return ok({ orderId: buyChannelData.order_id, channelOpenCost });
+	const feeRes = await updateFee({
```
unnecessary `await`

coreyphillips replied:
Update [here](https://github.com/synonymdev/bitkit/pull/1216/commits/e852886655563df0f4cf8f21e885e456b5cafb7d).

pwltr commented on src/store/actions/blocktank.ts:
```diff
@@ -265,19 +258,33 @@ export const startChannelPurchase = async ({
 	}
 
-	updateSendTransaction({
+	await updateSendTransaction({
```
unnecessary `await`

coreyphillips replied:
Update [here](https://github.com/synonymdev/bitkit/pull/1216/commits/e852886655563df0f4cf8f21e885e456b5cafb7d).

pwltr commented on src/store/actions/blocktank.ts:
```diff
@@ -98,11 +78,16 @@ export const refreshOrder = async (
-			const finalizeRes = await finalizeChannel(orderId);
+			const finalizeRes = await openChannel(orderId);
 			if (finalizeRes.isOk()) {
+				//setLightningSettingUpStep(2);
```
remove comment

pwltr commented on src/store/actions/blocktank.ts:
```diff
@@ -367,30 +375,31 @@ export const addPaidBlocktankOrder = ({
-
-	const oneOtherOrderHasState = (states: number[]): boolean => {
-		return otherOrderStates.some((state) => states.includes(state));
-	};
+	//TODO: Not sure what to do with this
+	// const currentOrders = getBlocktankStore().orders;
+	// const otherOrders = currentOrders.filter((o) => o.id !== order.id);
+	// const otherOrderStates = otherOrders.map((o) => o.state);
+	//
+	// const oneOtherOrderHasState = (states: number[]): boolean => {.
+	// 	return otherOrderStates.some((state) => states.includes(state));
+	// };
```
Can this be removed? Cards seem to work okay

coreyphillips replied:
Yes, that was sorted. Forgot to remove the todo. Updated [here](https://github.com/synonymdev/bitkit/pull/1216/commits/e852886655563df0f4cf8f21e885e456b5cafb7d).

pwltr reviewed (changes requested):
Couple of nitpicks other than that LGTM
